### PR TITLE
[WIP] Support for Frictionless datapackage

### PR DESCRIPTION
**Pull request by pathak-mayurdeep**
This is a separate branch where I have added some new datapackage to `datapackages.yml` and made some changes to debug the path issue while installing.

### CODE REVIEW

File: scripts/datapackages.yml [hunk 1]
```diff
@@ -1,3 +1,2 @@
-gdp: https://raw.githubusercontent.com/datasets/gdp/master/datapackage.json
-country-codes: https://raw.githubusercontent.com/datasets/country-codes/master/datapackage.json
-population-density: https://raw.githubusercontent.com/datasets/population/master/datapackage.json
+eu-emission-trading-system: https://datahub.io/core/eu-emissions-trading-system/r/eu-emissions-trading-system_zip.zip
```
Comment (henrykironde) [inline scripts/datapackages.yml]: https://pkgstore.datahub.io/core/eu-emissions-trading-system/13/datapackage.json

File: scripts/datapackages.yml [hunk 2]
```diff
@@ -1,3 +1,2 @@
...
+global_co2_levels : https://datahub.io/core/co2-ppm/r/co2-ppm_zip.zip
```
Comment (henrykironde) [inline scripts/datapackages.yml]: https://datahub.io/core/co2-ppm/datapackage.json

**Comment by pathak-mayurdeep** [comment]
I removed the path entirely and now I'm able to download the file. But there are problems while installing.

image

image

**Comment by henrykironde** [comment]
```bash
henry@retriever (test1420) $ retriever ls

 New additions

 New additions

 problem in converting

 problems creating mmodule --  local variable 'updated_resources' referenced before assignment
failed to load Frictionless Datapackage : country-codes (https://raw.githubusercontent.com/datasets/country-codes/master/datapackage.json)
Exception: local variable 'module' referenced before assignment

 New additions

 problem in converting

 problems creating mmodule --  local variable 'updated_resources' referenced before assignment
failed to load Frictionless Datapackage : population-density (https://raw.githubusercontent.com/datasets/population/master/datapackage.json)
Exception: local variable 'module' referenced before assignment
Available datasets : 224

abalone-age                             home-ranges                                  
acton-lake                              intertidal-abund-me                          
airports                                iris                                         
amniote-life-hist                       jornada-lter-rodent                          
antarctic-breed-bird                    la-selva-trees                               
aquatic-animal-excretion      
```

Looks like you are still reading and loading the old files. The first in the search path is used. In this case`scripts/datapackages.yml` will be read given that we are in the repo.


```bash
henry@retriever (test1420) $ retriever install postgres eu-emissions-trading-system

 New additions

 New additions
Creating database eu_emissions_trading_system...

Downloading be962818629bd84c103fbe68ae64a0f4.csv: 2.52kB [00:00, 3.98kB/s]                        
syntax error at or near "-"
LINE 1: CREATE TABLE eu_emissions_trading_system.eu-ets_csv (country...
                                                   ^
```

We should clean up the `JSON` value corresponding to the `name` in resources. Change it to lower and replace - with _

**Comment by pathak-mayurdeep** [comment]
@henrykironde sorry for such a late reply. I've been dealing with some health issues since last month, so I took some time off. I've recovered now and I would like to continue with this. 

Regarding the last issue about reading and loading the old files, could you elaborate. I'm not getting how it is still reading the old files.

**Comment by henrykironde** [comment]
I will test this again since it has been long since the last test

**Comment by pathak-mayurdeep** [comment]
@henrykironde Could you test this PR. Last time we had ran into some problems regarding reading the wrong `datapackages.yml`

File: retriever/lib/f_datapackage.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+     cleans the json file with the required changes, then creates script module 
+"""
+
+def get_dps():
+    """Load the list of data packages from datapackages.yml
+    Checks all of the search paths for a datapackages.yml file and loads it into
+    a dictionary.
+    """
+    for path in SCRIPT_SEARCH_PATHS:
+        try:
```
Comment (henrykironde) [inline retriever/lib/f_datapackage.py]: When you do a search for script(`SCRIPT_SEARCH_PATHS`). we start in the working directory, then `./scripts` until we get to the `~/.retriever/script directory`

As soon as we get the first script, in this case we look for `datapakages.yml`.  We should consider that as the most current file and jump out of the loop.

Comment (pathak-mayurdeep) [reply retriever/lib/f_datapackage.py]: I'm still trying to connect the dots here, this is taking forever..

What I understand is that `datapackages.yml` doesn't exit in `~/.retriever/scripts` ,
then I should probably try to solve that first.. yes?

I see that `SCRIPT_WRITE_PATH` points to `~/.retriever/scripts` and the function
`check_for_updates(repo=REPOSITORY)`  is being used to download the latest scripts to that path.

My thinking is I should create a similar function (or edit `check_for_updates`) for updating `datapackages.yml`. 
@henrykironde What are your views?

File: retriever/lib/f_datapackage.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            with open(os.path.join(path, "datapackages.yml"), 'r') as dp_file:
```
Comment (henrykironde) [inline retriever/lib/f_datapackage.py]: The solution to use the first script that you find would be to exit the for loop if you find the first file 

```python
def get_dps():
    """Load the list of data packages from datapackages.yml
    Checks all of the search paths for a datapackages.yml file and loads it into
    a dictionary.
    """
    for path in SCRIPT_SEARCH_PATHS:
        try:
            if os.path.isfile(os.path.join(path, "datapackages.yml")):
                with open(os.path.join(path, "datapackages.yml"), 'r') as dp_file:
                    dp_dict = yaml.safe_load(dp_file)
                return dp_dict
        except:
            pass
    return None
```
With this method, developers will always use the file in the working directory/scripts and users will always use ~/.retriever/scripts  as the defaults.

**Comment by henrykironde** [comment]
Could you paste the commands that you have run and outputs. Also if you are done remove the `WIP` tag

**Comment by pathak-mayurdeep** [comment]
image

**Comment by pathak-mayurdeep** [comment]
@henrykironde I think this issue is fixed. I had to do some cleanup of the table names and a few other things.
image

image

I'll test some more by adding other links from data hub awesome. If all goes correctly, then the original issue #1404 will hopefully be resolved.

**Comment by henrykironde** [comment]
What does the value of retriever ls give you?

**Comment by pathak-mayurdeep** [comment]
@henrykironde 
image
image

**Comment by pathak-mayurdeep** [comment]
This PR was only for testing. I'll do a cleanup of the whole code and create a separate PR after I'm done testing everything. 
Also, should I add all the links you mentioned in #1404 or do you want it done later @henrykironde?

**Comment by henrykironde** [comment]
> Also, should I add all the links you mentioned in #1404 or do you want it done late

We can do that later

**Comment by henrykironde** [comment]
```j
✦2 ❯ retriever install postgres co2-ppm

 New additions

 New additions
Creating database co2_ppm...

Downloading 8caf50a94455f329a5d258cd2b6f5f03.csv: 15.0B [00:00, 68.7B/s]                                           
syntax error at or near "-"
LINE 1: CREATE TABLE co2_ppm.co2-mm-mlo_csv (record_date varchar, de...
                                ^

Replacing existing table
Downloading 0b618d8b29775ea68c1a424f24f5dd88.csv: 1.00B [00:00, 3.74B/s]
syntax error at or near "-"
LINE 1: CREATE TABLE co2_ppm.co2-annmean-mlo_csv (year varchar, mean...
                                ^

Replacing existing table
Downloading 6faf34070c171575a4cb72ce0b56c17b.csv: 1.00B [00:00, 3.95B/s]
syntax error at or near "-"
LINE 1: CREATE TABLE co2_ppm.co2-gr-mlo_csv (year varchar, annual_in...
                                ^

Replacing existing table
Downloading 5fa2d3229917fad2b943705cee6f09eb.csv: 8.00B [00:00, 32.2B/s]                                           
syntax error at or near "-"
LINE 1: CREATE TABLE co2_ppm.co2-mm-gl_csv (record_date varchar, dec...
                                ^

Replacing existing table
Downloading cf3ab8d8771c2bfea077adaa9a853395.csv: 1.00B [00:00, 6.10B/s]
syntax error at or near "-"
LINE 1: CREATE TABLE co2_ppm.co2-annmean-gl_csv (year varchar, mean ...
                                ^

Replacing existing table
Downloading aff4f775ed0d71fa76b2ec75d8ca7db9.csv: 1.00B [00:00, 5.57B/s]
syntax error at or near "-"
LINE 1: CREATE TABLE co2_ppm.co2-gr-gl_csv (year varchar, annual_inc...
                                ^

Replacing existing table
syntax error at or near "-"
LINE 1: CREATE TABLE co2_ppm.co2-mm-mlo (record_date varchar, decima...
                                ^

Replacing existing table
syntax error at or near "-"
LINE 1: CREATE TABLE co2_ppm.co2-annmean-mlo (year varchar, mean dou...
                                ^

Replacing existing table
syntax error at or near "-"
LINE 1: CREATE TABLE co2_ppm.co2-gr-mlo (year varchar, annual_increa...
                                ^

Replacing existing table
syntax error at or near "-"
LINE 1: CREATE TABLE co2_ppm.co2-mm-gl (record_date varchar, decimal...
                                ^

Replacing existing table
syntax error at or near "-"
LINE 1: CREATE TABLE co2_ppm.co2-annmean-gl (year varchar, mean doub...
                                ^

Replacing existing table
syntax error at or near "-"
LINE 1: CREATE TABLE co2_ppm.co2-gr-gl (year varchar, annual_increas...
                                ^

Replacing existing table
```

```j
retriever on  master [⇡?] via 🐍 v3.7.6 took 7s 
✦2 ❯ retriever install postgres eu-emissions-trading-system

 New additions

 New additions
Creating database eu_emissions_trading_system...

schema "eu_emissions_trading_system" already exists

Installing into existing database
Downloading be962818629bd84c103fbe68ae64a0f4.csv: 2.52kB [00:00, 3.57kB/s]                                         
syntax error at or near "-"
LINE 1: CREATE TABLE eu_emissions_trading_system.eu-ets_csv (country...
                                                   ^

Replacing existing table
syntax error at or near "-"
LINE 1: CREATE TABLE eu_emissions_trading_system.eu-ets (country_cod...
                                                   ^

Replacing existing table
```
retriever ls runs fine 
I have tried to install the dataset but there is an error.
In the function `def convert_frictionless` we need to replace "-" in the resource name value.

**Comment by pathak-mayurdeep** [comment]
@henrykironde actually I haven't pushed the latest commits yet. I'll do that tomorrow morning (IST).

**Comment by henrykironde** [comment]
@pathak-mayurdeep get time and we get this tested together.

**Comment by ethanwhite** [comment]
Replaced by #1489

**Pull request closed without merging**